pandas/io/xml.py

```diff
@@ -0,0 +1,1017 @@
+                self.xpath + "[1]/*", namespaces=self.namespaces
+            )
+
+            if is_list_like(self.names):
+                if len(self.names) < len(children):
+                    raise ValueError(
+                        "names does not match length of child elements in xpath."
+                    )
+            else:
+                raise TypeError(
+                    f"{type(self.names).__name__} is not a valid type for names"
+                )
+
+    def _convert_io(self) -> Union[None, str]:
+        """
+        Convert filepath_or_buffer object to string.
+
+        This method will convert io object into a string or keep
+        as string, depending on object type.
+        """
+
+        obj = None
+
+        if isinstance(self.raw_doc, str):
+            obj = self.raw_doc
+
+        if isinstance(self.raw_doc, bytes):
+            obj = self.raw_doc.decode(self.encoding)
+
+        if isinstance(self.raw_doc, io.StringIO):
+            obj = self.raw_doc.getvalue()
+
+        if isinstance(self.raw_doc, io.BytesIO):
+            obj = self.raw_doc.getvalue().decode(self.encoding)
+
+        if isinstance(self.raw_doc, io.TextIOWrapper):
+            obj = self.raw_doc.read()
+
+        if isinstance(self.raw_doc, io.BufferedReader):
+            obj = self.raw_doc.read().decode(self.encoding)
+
+        return obj
+
+    def _parse_doc(self):
+        """
+        Build tree from io.
+
+        This method will parse io object into tree for parsing
+        conditionally by its specific object type.
+
+        Raises
+        ------
+        HttpError
+            * If URL cannot be reached.
+
+        LookupError
+            * If xml document has incorrect or unknown encoding.
+
+        OSError
+            * If file cannot be found.
+
+        XMLSyntaxError
+            * If xml document conntains syntax issues.
+
+        ValueError
+            * If io object is not readable as string or file-like object.
+        """
+
+        from lxml.etree import XML, XMLParser, XMLSyntaxError, parse
+
+        self.raw_doc = self.stylesheet if self.is_style else self.io
+
+        current_doc = self._convert_io()
+        if current_doc:
+            is_xml = current_doc.startswith(("<?xml", "<"))
+        else:
+            raise ValueError("io is not a url, file, or xml string")
+
+        try:
+            curr_parser = XMLParser(encoding=self.encoding)
+
+            if is_url(current_doc):
+                with urlopen(current_doc) as f:
+                    r = parse(f, parser=curr_parser)
+            elif is_xml:
+                r = XML(bytes(current_doc, encoding=self.encoding))
+            else:
+                r = parse(current_doc, parser=curr_parser)
+        except (LookupError, URLError, HTTPError, OSError, XMLSyntaxError) as e:
+            raise e
+
+        return r
+
+
+def _data_to_frame(data, **kwargs) -> DataFrame:
+    """
+    Convert parsed data to Data Frame.
```

Comment:
rather than that putting the writing in pandas/io/xml.py and the reading in the formatters I would be tempted to put *all* of this in pandas/io/xml/reader.py and writer.py and have base.py to shared things (or methods if needed). just so all of its in the same place.

Reply:
also can centralize the doc-strings rather than repeating.

Reply:
I believe it is the other way around. I followed `html` setup, so `pandas/io/xml.py` is the reader and `pandas/io/formats/xml.py` the writer script. And like `to_csv, `to_html`, `to_string`, and few others, their formatter classes are called in `pandas/io/formats/formats.py` which somewhat extends the formatter class is initially called in `pandas/core/frame.py`. And `frame.py` and `formats.py` pass through same args, hence repetition. I think `html` and `string` share similar docstrings. At first, I was confused by this setup. To centralize may be substantial handling of scripts. Maybe an i/o code clean up PR?